src/v/cloud_storage/remote_segment_index.cc

```diff
@@ -184,6 +184,58 @@ offset_index::find_kaf_offset(kafka::offset upper_bound) {
     return res;
 }
 
+std::optional<offset_index::find_result>
+offset_index::find_timestamp(model::timestamp upper_bound) {
+    if (_initial_time == model::timestamp::missing()) {
+        // Bail out early if this is a version 1 index
+        // that does not index timestamps.
+        return std::nullopt;
+    }
+
+    size_t ix = 0;
+
+    auto search_result = maybe_find_offset(
+      upper_bound.value(), _time_index, _time_offsets);
+
+    if (std::holds_alternative<std::monostate>(search_result)) {
+        return std::nullopt;
+    } else if (std::holds_alternative<find_result>(search_result)) {
+        return std::get<find_result>(search_result);
+    }
+    auto maybe_ix = std::get<index_value>(search_result);
```

Comment:
ideally we could use `ss::visit` so we get type checking on the variant, otherwise, should we add a holds alternative assertion here to guard against future additions to the variant types?

Reply:
Agreed. Added a new commit which refactors to use `ss::visit`.